numpy/_core/src/multiarray/ctors.c

```diff
@@ -2207,43 +2207,47 @@ PyArray_FromInterface(PyObject *origin)
      * the 'descr' attribute.
      */
     if (dtype->type_num == NPY_VOID) {
-        PyObject *descr = _PyDict_GetItemStringWithError(iface, "descr");
-        if (descr == NULL && PyErr_Occurred()) {
+        PyObject *descr = NULL;
+        result = PyDict_GetItemStringRef(iface, "descr", &descr);
+        if (result == -1) {
             goto fail;
         }
         PyArray_Descr *new_dtype = NULL;
-        if (descr != NULL) {
+        if (result == 1) {
             int is_default = _is_default_descr(descr, attr);
             if (is_default < 0) {
+                Py_DECREF(descr);
                 goto fail;
             }
             if (!is_default) {
                 if (PyArray_DescrConverter2(descr, &new_dtype) != NPY_SUCCEED) {
+                    Py_DECREF(descr);
                     goto fail;
                 }
                 if (new_dtype != NULL) {
-                    Py_DECREF(dtype);
-                    dtype = new_dtype;
+                    Py_SETREF(dtype, new_dtype);
                 }
             }
-
         }
-
+        Py_DECREF(descr);
```

Comment:
@ngoldbaum this is at the wrong indentation level.  It's too late here, and I am not sure where to put a test though.  `cunumeric` notices this, it seems they use void (I suspect unstructured but dunno), without a `descr`.

Reply:
Oops, thanks for catching. I unfortunately can't help with a test, I was doing a mechanical refactor here.

Reply:
We must have some `__array_interface__` tests.  It shouldn't be hard to create a dict that runs into this, tbh.  Although I could overlook it, it would be nice to just have.

Reply:
My initial reading of your message was that it was a head's up and an ask for help. I'm happy to fix this but if, in the future you want me to look at something, it would help me if you could say that explicitly in your initial message to me.

Reply:
I don't think the indentation level is wrong, doesn't it just need to be an xdecref?

I also don't immediately see how to write a test for this, the existing `__array_interface__` tests mostly use the ndarray `__array_interface__` implementation. It would be a good idea to improve the tests there but I don't want to do that just to fix this bug.

Reply:
Should have just created an issue.  I wanted to look at it quickly, but realized that adding a test was not quite as quick as I had hoped, and hoped you might have a look.